internal/gengapic/gengapic.go

```diff
@@ -318,56 +318,70 @@ func (g *generator) insertDynamicRequestHeaders(m *descriptor.MethodDescriptorPr
 		return err
 	}
 	if len(headers) > 0 {
-		g.printf("routingHeaders := \"\"")
+		g.printf(`routingHeaders := ""`)
 		g.printf("routingHeadersMap := make(map[string]string)")
 		for i, h := range headers {
+			namedCaptureRegex := h[0]
 			field := h[1]
 			headerName := h[2]
-			g.printf("routingRegexp%v := regexp.MustCompile(\"%s\")", i, h[0])
 			accessor := fmt.Sprintf("req%s", fieldGetter(field))
-			regexHelper := fmt.Sprintf("routingRegexp%v.FindStringSubmatch(%s)[1]", i, accessor)
-			g.printf("headerName%v := \"%s\"", i, headerName)
-			g.printf("headerValue%v := \"\"", i)
-
+			regexHelper := fmt.Sprintf("reg.FindStringSubmatch(%s)[1]", accessor)
 			regexHelper, err = g.getFormattedValue(m, field, regexHelper)
 			if err != nil {
 				return err
 			}
-			g.printf("if routingRegexp%v.MatchString(%s){headerValue%v = %s}", i, accessor, i, regexHelper)
-			// This enables overwriting of headerNames such that the last annotation to match will be the value sent.
-			g.printf("if len(headerValue%v) > 0{routingHeadersMap[headerName%v] = headerValue%v}", i, i, i)
+			g.printf("if reg := regexp.MustCompile(\"%s\");", namedCaptureRegex)
+			g.printf("reg.MatchString(%s) {", accessor)
+			g.printf("routingHeadersMap[\"%s\"] = %s", headerName, regexHelper)
+			// If there are multiple rules for the same header name, the last rule that matches wins.
+			if i+1 < len(headers) && headers[i+1][2] == headers[i][2] {
+				namedCaptureRegex := headers[i+1][0]
+				field := headers[i+1][1]
+				accessor := fmt.Sprintf("req%s", fieldGetter(field))
+				regexHelper := fmt.Sprintf("reg.FindStringSubmatch(%s)[1]", accessor)
+				regexHelper, err = g.getFormattedValue(m, field, regexHelper)
+				if err != nil {
+					return err
+				}
+				g.printf("} else if reg = regexp.MustCompile(\"%s\");", namedCaptureRegex)
+				g.printf("reg.MatchString(%s) {", accessor)
+				g.printf("routingHeadersMap[\"%s\"] = %s", headerName, regexHelper)
+				g.printf("}")
+			} else {
+				g.printf("}")
+			}
 		}
-		g.printf("for headerName, headerValue := range routingHeadersMap {routingHeaders = routingHeaders + fmt.Sprintf(\"%%s=%%s&\", headerName, headerValue)}")
-		// Trim the trailing ampersand symbol.
-		g.printf("if len(routingHeaders) > 0{routingHeaders = routingHeaders[:len(routingHeaders)-1]}")
+		g.printf("for headerName, headerValue := range routingHeadersMap {")
+		g.printf("routingHeaders = routingHeaders + fmt.Sprintf(\"%%s=%%s&\", headerName, headerValue)")
```

Comment:
indent, use backtick so you don't need to escape the double quotes, concat with `routingHeaders` by including it in the format string with leading `%s`
```suggestion
		g.printf(`  routingHeaders = fmt.Sprintf("%%s%%s=%%s&", routingHeaders, headerName, headerValue)`)
```

Reply:
done